sw/device/tests/crypto/kmac_verify_functest.c

```diff
@@ -0,0 +1,71 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/crypto/drivers/kmac.h"
+#include "sw/device/lib/crypto/include/datatypes.h"
+#include "sw/device/lib/crypto/include/hash.h"
+#include "sw/device/lib/crypto/include/mac.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+// The autogen rule that creates this header creates it in a directory named
+// after the rule, then manipulates the include path in the
+// cc_compilation_context to include that directory, so the compiler will find
+// the version of this file matching the Bazel rule under test.
+#include "kmac_verify_testvectors.h"
+
+OTTF_DEFINE_TEST_CONFIG();
+bool test_main(void) {
+  LOG_INFO("Testing cryptolib KMAC driver.");
+
+  // Initialize the core with default parameters
+  kmac_error_t err = kmac_hwip_default_configure();
+  if (err != kKmacOk) {
+    return err;
+  }
+
+  for (size_t i = 0; i < ARRAYSIZE(nist_kmac_vectors); i++) {
+    kmac_test_vector_t *current_test_vector = &nist_kmac_vectors[i];
+    LOG_INFO("loop counter = %d, vector identifier: %s", i,
+             current_test_vector->vector_identifier);
+
+    // The test code supports digest sizes up to `MAX_DIGEST_SIZE` bytes.
+    // `MAX_DIGEST_SIZE` comes from generated header file.
+    uint8_t digest[MAX_DIGEST_SIZE];
```

Comment:
Thanks Jade, I didn't know variable length arrays were allowed in C. Since the size is not known during compile time, who allocates this array? Which memory region it goes to, is there something like a heap?

(+ made this change and pushed again)

Reply:
It's allowed in C but not always a good idea because the compiled code will be less straightforward. However, in this case, since the code is test-only and there's some significant complexity in calculating the maximum size anyway, it seemed to make sense. Your instincts are good, though -- it's often not a good idea in performance- or security-critical code :slightly_smiling_face: 

I'm not sure which memory region, but my mental model is that the array gets allocated at runtime and then freed when the variable goes out of scope; it's got special compiler support iirc.